BJS2-15564 - "Организация события"

i7rincpane opened the pull request:
Реализовал возможность создавать событие.

Pa4aPv committed:
BJS2-15564 - "Организация события"
Реализовал возможность создавать событие.

Pa4aPv committed:
BJS2-15564 - "Организация события"
Реализовал получение всех событий с фильтрами

Pa4aPv committed:
BJS2-15564 - "Организация события"
Реализовал получение события

Pa4aPv committed:
BJS2-15564 - "Организация события"
Реализовал удаление события

Pa4aPv committed:
BJS2-15564 - "Организация события"
Реализовал получение всех созданные пользователем событий

Pa4aPv committed:
BJS2-15564 - "Организация события"
Реализовал получение всех событий, в которых пользователь принимает участие

Pa4aPv committed:
BJS2-15564 - "Организация события"
Отредактировал метод получение по фильтру на реализацию через стрим + динамическое добавление фильтров + протестировал реализацию

pablo023 commented on src/main/java/school/faang/user_service/controller/event/EventController.java:
```diff
@@ -0,0 +1,52 @@
+                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND));
+    }
+
+    public void delete(Long id) {
+        if (!eventService.delete(id)) {
+            throw new ResponseStatusException(HttpStatus.NOT_FOUND);
+        }
+    }
```
Чет ты тут не дописал)

i7rincpane replied:
подскажи что, не пойму) я с рестом мало работал, в mvc редиректил на вьюшки .. тут тоже типо редирект нужен, или что?

pablo023 commented on src/main/java/school/faang/user_service/controller/event/EventController.java:
Смотри, разделение обязанностей это ключ. Контроллеры только проверяют валидность данных, через аннотацию Valid. Всю серьёзную работу по обеспечению порядка должен делать сервисный слой. Там всё и происходит: проверки, валидации ID и так далее. Это удобно, потому что ты можешь один и тот же код использовать в разных местах, не копируя его туда-сюда. Также это упрощает поддержку и обновление приложения — меняешь что-то в одном месте, а не бегаешь по всем контроллерам. Тестировать тоже легче, потому что тестируешь логику отдельно от веб-мороки. И безопаснее это, потому что ты уверен, что валидации сработают всегда, независимо от того, через что пользователь пришёл. Ну и наконец, так проще масштабировать приложение — добавляешь новые функции, расширяешься, и всё это без боли и страданий. Вот так вот все дело. Иногда в контроллерах происходит логика по работе с запросом - там контекст есть, и из него бывает нужно сформировать какие-то данные, которые отправляются на сервис. Но валидации, и исключения, которые летят с сервиса - оставь сервису.

i7rincpane replied:
хороший комментарий, побольше таких надо) но вроде бы я так и сделал? уточни если что то забыл

pablo023 replied:
Будет тебе побольше таких, не проблема) Зачем я его написал - по методу delete. Лучше вынести валидацию во всех её видах, в сервисный слой. Это даже выглядеть будет проще и чище. А оттуда уже кидай валидаторами что хочешь)

pablo023 commented on src/main/java/school/faang/user_service/dto/SkillReadDto.java:
```diff
@@ -0,0 +1,7 @@
+
+import lombok.Value;
+
+@Value
+public class SkillReadDto {
+}
```
Чет тут не дописал

i7rincpane replied:
не совсем понимаю нужно ли мне его заполнять, это вроде как не моя задача и он уже при мердже получается будет кем-то заполнен? или мне нужно найти задачу для скилов и заполнить в соответствии с её требованиями?

pablo023 replied:
Если у тебя возникает ситуация, что тебе нужен какой-то класс из другой задачи, то у тебя есть такие варианты:

1) Сделать заглушку
2) Пойти к коллеге, который занимается этой функциональностью, и узнать у него статус. Если он уже написал этот код - можно попросить его этот код, например, либо подождать, пока его код будет вмерджен
3) Если на текущий момент задача, от которой зависит твой код - ни на ком, ты можешь ознакомиться с её условием, и реализовать какую-то часть функционала.

Но так или иначе, я не понял сути этого пустого класса - ты ведь нигде не применяешь его. Зачем он тебе? Я подумал, он тебе нужен, и ты его забыл заполнить, а он.. Не нужен)

pablo023 commented on src/main/java/school/faang/user_service/dto/event/EventCreateEditDto.java:
```diff
@@ -0,0 +1,23 @@
+@Value
+public class EventCreateEditDto  {
+
+    String title;
+    LocalDateTime startDate;
+    LocalDateTime endDate;
+    Long ownerId;
+    String description;
+    List<Long> relatedSkillIds;
+    String location;
+    int maxAttendees;
+    EventType eventType;
+    EventStatus eventStatus;
```
Чет модификаторы доступа забыл)

i7rincpane replied:
аннотация @value делает класс неизменяемым - добавляет модификатор private

pablo023 commented on src/main/java/school/faang/user_service/dto/event/EventReadDto.java:
```diff
@@ -0,0 +1,19 @@
+@Value
+public class EventReadDto {
+
+    Long id;
+    String title;
+    LocalDateTime startDate;
+    LocalDateTime endDate;
+    UserReadDto owner;
+    String description;
+    String location;
+    int maxAttendees;
```
Тоже забыл модификаторы)

i7rincpane replied:
да не забыл)

pablo023 commented on src/main/java/school/faang/user_service/mapper/Mapper.java:
Маппинг делаем через MapStruct - это либа отдельная)

i7rincpane replied:
жду решения по этому вопросу, нужно ли переписывать

pablo023 commented on src/main/java/school/faang/user_service/mapper/UserReadMapper.java:
Про MapStruct не увидел в таске)

i7rincpane replied:
увидел, но боюсь её) + хочу контролировать процесс маппинга и иметь возможность добавлять сложную логику с подключением различных вложенностей

pablo023 commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,90 @@
+                .map(readMapper::map);
+    }
+
+    private void validate(EventCreateEditDto eventDto) {
+        var validationResult = validator.validate(eventDto);
+        if (validationResult.hasErrors()) {
+            throw new DataValidationException(validationResult.getErrors());
+        }
+    }
```
Лучше бы валидацию вынести в отдельный класс валидатор, т.к. они у нас могут добавляться существенно по мере роста приложения.

i7rincpane replied:
жду решения по этому вопросу, нужно ли на каждый не валидный случай выкидывать исключение, или оставить реализацию где они копятся и выкидываются кучей.. как понял, вопорос в том могут ли два пользователя использовать один экземпляр сервиса и конкурировать за использование валидатора и совместно копить эти ошибки ?

pablo023 commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,90 @@
+
+    public Optional<EventReadDto> findById(Long id) {
+        return repository.findById(id)
+                .map(eventReadMapper::map);
```
Ну и не просто мап, а toDto это будет)

i7rincpane replied:
принял

pablo023 reviewed (changes requested):
Поторопился, наверное, Паш) Надо переделать чуток)

pablo023 commented on src/main/java/school/faang/user_service/mapper/UserReadMapper.java:
```diff
@@ -0,0 +1,14 @@
+    public UserReadDto map(User object) {
+        return new UserReadDto(object.getId());
+    }
+}
```
Ну Паш, это же неудобно получается. Ты вот написал логику маппинга здесь. Добавишь поле - тебе сюда идти и переписывать билдером как-то это всё придется)

pablo023 commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    private void validate(EventCreateEditDto eventDto) {
+        var validationResult = validator.validate(eventDto);
```
Вот этот кейс с варом не особо уместен. Чтобы понять, что validate валидатора возвращает, нужно в него идти.

pablo023 commented on src/main/java/school/faang/user_service/valitator/Error.java:
В Java уже есть класс Error - подкласс Throwable. Когда ты создаешь такой класс для своих каких-то задач, ты вносишь путаницу в кодовую базу. Только ты будешь знать о том, что вот этот Error - это твой класс. Если ты на рабочем проекте будешь применять такую логику, то ты создашь проблемы тем, кто придет туда через несколько лет, и будет разбираться - что еще за Error тут кастомный.

pablo023 commented on src/main/java/school/faang/user_service/valitator/ValidationResult.java:
Смотри, как у нас это будет работать. При работе приложения, будет какая-то стратегия создания этого класса. В любом случае, это сведется к трем вариантам:

1) Это будет синглтон. В таком случае, у тебя спринг-приложение будет рвать вот этот вот список, и набивать его хаотично, и лететь у тебя будет непойми что из него.

2) Он будет создваться каждый раз новый, для каждого потока. Для того, чтобы создать класс, и коллекцию вот эту - ты будешь затрачивать как процессорное время, так и память. Вопрос - для чего, какую проблему решает такой подход, жертвуя памятью и процессором?

3) Он будет создаваться и убиваться по вызову. Опять таки - для чего?) Если у тебя будет валидатор, который принимает параметры, валидирует их, и кидает эксепшен - у тебя не будет никаких жертв по памяти вообще, и  у тебя будет простой булеан метод.

pablo023 commented on src/test/java/school/faang/user_service/mapper/event/EventCreateEditMapperTest.java:
```diff
@@ -0,0 +1,98 @@
+        assertThat(actualResult).isEqualTo(expectedResult);
+    }
+
+    private Event getEvent() {
+        return Event.builder()
+                .title("title")
+                .startDate(DATE_NOW)
+                .endDate(DATE_NOW.plusDays(1))
+                .owner(getUser(1L))
+                .description("description")
+                .relatedSkills(Stream.of(1L, 2L)
+                        .map(this::getSkill)
+                        .toList())
+                .type(EventType.POLL)
+                .status(EventStatus.PLANNED)
+                .maxAttendees(1)
+                .build();
+    }
```
Глянь аннотации BeforeEach, и BeforeAll

pablo023 commented on src/main/java/school/faang/user_service/controller/event/EventController.java:
Глянь ещё для контроллера интересные аннотации RequiredMapping, и ResponseStatus. Сейчас это необязательно, но я думаю, тебе будет интересно.

pablo023 reviewed (changes requested):
Смотри, Паш, у тебя сейчас зона роста - это понимание абстракции, и её цели. Я не знаю зачем ты в целях написал, что ты хочешь джун+, или что-то такое, тебе до хорошего мидла тут осталось ничего. Давай определим хорошего мидла - одна из самых проблемных зон у мидлов, это - абстракции. У тебя сейчас есть понимание - что это такое, но нет понимания - для чего она нужна. Абстракция нужна там, где с помощью неё, ты усложнишь прямо сейчас жизнь себе, написав больше кода, но упростишь жизнь всем остальным в дальнейшем, потому что твой код будет чище, его будет легче читать, его будет легче расширять, и с ним будет проще работать другим разработчикам. На данный момент ты используешь абстракцию там, где ты видишь, что её можно использовать, просто так. Всегда, при любых обстоятельствах, твой код должен стремиться к тому, чтобы быть минималистичным, и максимально простым. Когда ты хочешь написать что-то, что угодно - подумай, как это сделать максимально просто. Когда ты понял - подумай, принесет ли такой подход проблемы прямо сейчас, либо в дальнейшем. Если принесет - подумай над тем, как сделать красиво. В любом случае, рабочий процесс в командой разработке - это условия коротких сроков, и код-ревью - это часть этого процесса. Когда ты пишешь коротко, красиво, и по делу - тебя быстро и легко ревьюить, и ты будешь получать быстрые аппрувы от старших коллег. Если ревью будет идти непросто - твой код будут либо возвращать, либо ревьюить сквозь пальцы, а потом, когда всё сломается - скандалить, что ну а как там ревьюить было. Если ты хочешь творческого простора, и прокачать свои скиллы - попроси сложную задачу, у нас есть реально сложные вещи, на которых дофига чему можно научиться. Но не усложняй простое ;-) 

Насчёт маппера - сделай с мапстрактом. Не бойся его, если надо будет - объясним. Тебе нужно, чтобы тебе было комфортно работать с любой технологией. Мапстракт, кстати, отличная. Но комфортно тебе должно быть с чем угодно, т.к. задачи - это задачи, и хоть на JS напишешь, если надо будет. Если у тебя есть желание реализовать свой мапстракт чисто из интереса - я не против, можем это отдельно обсудить.

Pa4aPv committed:
BJS2-15564 - "Организация события"
1. Переписал кастомные маперы на мапстракт
2. Изменил валидатор с проверки листа ошибок
3. На штучную проверку, изменил метод удаления

pablo023 commented on src/main/java/school/faang/user_service/controller/event/EventController.java:
```diff
@@ -0,0 +1,50 @@
+                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND));
+    }
+
+    public void delete(Long id) {
```
deleteEvent*

pablo023 commented on src/main/java/school/faang/user_service/controller/event/EventController.java:
```diff
@@ -0,0 +1,50 @@
+        eventService.delete(id);
+    }
+
+    public EventReadDto update(Long id, EventCreateEditDto editDto) {
```
updateEvent

pablo023 commented on src/main/java/school/faang/user_service/controller/event/EventController.java:
```diff
@@ -0,0 +1,50 @@
+                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND));
+    }
+
+    public List<EventReadDto> getOwnedEvents(long userId) {
```
getEventsByUserId

pablo023 commented on src/main/java/school/faang/user_service/dto/event/EventCreateEditDto.java:
Если ты хочешь разделить входные и выходные, то лучше ReadEvetDto / WriteEventDto. Это общая практика, так чаще всего называют.

pablo023 commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,84 @@
+
+    private final EventRepository repository;
+    private final EventCreateEditMapper createEditMapper;
+    private final EventReadMapper readMapper;
```
eventToReadEventDtoMapper* С верхним аналогично.

pablo023 commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,84 @@
+        return readMapper.map(event);
+    }
+
+    public List<EventReadDto> findAllBy(EventFilterDto filter) {
```
by what?

pablo023 commented on src/main/java/school/faang/user_service/filter/event/EventFilter.java:
Упусил isApplicable

pablo023 commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,84 @@
+                .toList();
+    }
+
+    public Optional<EventReadDto> findById(Long id) {
+        return repository.findById(id)
+                .map(readMapper::map);
+    }
```
В этом методе у тебя бизнес-логика порвалась на контроллер и сервис. Пускай тут хэндлинг опшинала будет, и возвращаеся либо дто, либо исключение.

pablo023 commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,84 @@
+
+    @Transactional
+    public void delete(Long id) {
+        //TODO:12.07.2024 не понимаю, NOT_FOUND это же не совсем валидация? посути это запрос
+        // на не существующий контент и это ведь обязанность контроллера выкинуть HttpStatus.NOT_FOUND? или всетаки
+        // пусть сервайс ошибку валидации выкидывает?
```
Эту ошибку пусть кидает сервис. Контроллер может кидать только ошибки, связанные с валидацией данных, такие как на нулл, либо несоотвествие типов. Когда ожидал лонг или инт, а получил строку, и тд.

pablo023 commented on src/main/java/school/faang/user_service/mapper/event/EventCreateEditMapper.java:
```diff
@@ -0,0 +1,45 @@
+            @Mapping(target = "owner", expression = "java(findUserById(eventCreateEditDto.getOwnerId()))"),
+            @Mapping(target = "relatedSkills", expression = "java(findSkillsByIds(eventCreateEditDto.getRelatedSkillIds()))")
+    })
+    public abstract Event map(EventCreateEditDto eventCreateEditDto, @MappingTarget Event event);
```
А зачем так? Либа позволяет писать просто:

Event toEvent(EventDto eventDto) - и всё. Это всё, что мапстракту нужно)

pablo023 commented on src/main/java/school/faang/user_service/valitator/event/EventCreateEditValidator.java:
```diff
@@ -0,0 +1,43 @@
+
+    private final SkillRepository skillRepository;
+
+    public void validate(EventCreateEditDto object) {
+        if (object.getStartDate() == null) {
+            throw new DataValidationException(getNotNullErrorText("start-date"));
+        }
+        if (StringUtils.isEmpty(object.getTitle())) {
+            throw new DataValidationException(getNotNullErrorText("title"));
+        }
+        for (Long relatedSkillId : object.getRelatedSkillIds()) {
+            if (!getUserSkillIds(object.getOwnerId()).contains(relatedSkillId)) {
+                throw new DataValidationException(String.format("Не возможно установить skill, id: %s", relatedSkillId));
+            }
+        }
+    }
```
Лучше не создавать god-method, который как будто бы всё прям делает. Логика будет расти, фичи добавляться, и этот метод разжиреет до нечитаемости.

pablo023 reviewed (changes requested):
Исправь мелочевку, и мержим)

Pa4aPv committed:
BJS2-15564 - "Организация события"
1. переименовал маперы и дто
2. переминовал методы контроллера
3. валидатор расписал на отдельные методы
4. перенес обработку not found исключения из контроллера в сервис
5. декомпозировал логику фильтра отвечающий за применимость, в отдельный метод

pablo023 commented on src/main/java/school/faang/user_service/mapper/event/WriteEventDtoToEventMapper.java:
```diff
@@ -0,0 +1,45 @@
+    @Autowired
+    protected SkillRepository skillRepository;
+
+    @Mappings({
```
Mappings не обязательно, можешь просто вешать один маппинг за другим.

pablo023 commented on src/main/java/school/faang/user_service/mapper/event/WriteEventDtoToEventMapper.java:
```diff
@@ -0,0 +1,45 @@
+    })
+    public abstract Event map(WriteEventDto writeEventDto, @MappingTarget Event event);
+
+    protected User findUserById(Long id) {
+        return userRepository.findById(id).orElse(null);
+    }
```
Не надо нуллы кидать вообще нигде и вообще никогда. То, что нулл в джаве есть - это антипаттерн на уровне языка программирования вообще. Но коль уж он есть, всегда избегай того, чтобы он откуда-то возвращался. Хэндли ситуации с нуллом, не возвращай его.

pablo023 commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,82 @@
+    public ReadEvetDto findById(Long id) {
+        return repository.findById(id)
+                .map(eventToReadEventDtoMapper::map)
+                .orElseThrow(() -> new DataValidationException("Event not found"));
```
Такое сообщение ни о чём не говорит, никак не поможет найти ошибку и исправить её. Хотя бы нужен айдишник. Я бы сюда добавил еще логгер, и написал бы что-то типа такого

log.error(что не нашел, с какими данными, в каком сервисе, в каком методе)

pablo023 commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,82 @@
+        return eventToReadEventDtoMapper.map(event);
+    }
+
+    public List<ReadEvetDto> findAllByFilter(EventFilterDto eventFilterDto) {
```
Transactional, readOnly = true

pablo023 commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,82 @@
+
+    @Transactional
+    public void delete(Long id) {
+        Event entity = repository.findById(id).orElseThrow(() -> new DataValidationException("Event not found"));
```
Какой эвент не нашелся?

pablo023 commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,82 @@
+@RequiredArgsConstructor
+@Transactional(readOnly = true)
+public class EventService {
+    private final EventRepository repository;
```
Почему просто repository? У нас когда-то может появиться тут еще один репозиторий, и надо будет идти править в каждом месте, чтобы разграничивать и понимать с каким работаем? Идея же сама предлагает назвать такой объект eventRepository

pablo023 commented on src/main/java/school/faang/user_service/valitator/event/WriteEventValidator.java:
```diff
@@ -0,0 +1,56 @@
+            }
+        }
+    }
+
```
checkTitle*

pablo023 commented on src/main/java/school/faang/user_service/valitator/event/WriteEventValidator.java:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    private String getNotNullErrorText(String fieldName) {
+        return fieldName + " не может быть пустым";
```
Let's try to use more of English, ok?

Pa4aPv committed:
BJS2-15564 - "Организация события"
Исправил нейминг, добавил информацию об ошибке в исключение + логер, убрал mappings аннотацию, обработал нул ситуацию в дто мапере

pablo023 reviewed (approved):
Окей, мержим)
